Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.cs

```diff
@@ -0,0 +1,395 @@
+
+        if (userGun != gun) return;
+
+        StopShooting(gun);
+    }
+
+    private void OnGetState(EntityUid uid, GunComponent component, ref ComponentGetState args)
+    {
+        args.State = new NewGunComponentState
+        {
+            NextFire = component.NextFire,
+            ShotCounter = component.ShotCounter,
+            FakeAmmo = component.FakeAmmo,
+            SelectiveFire = component.SelectedMode,
+            AvailableSelectiveFire = component.AvailableModes,
+        };
+    }
+
+    private void OnHandleState(EntityUid uid, GunComponent component, ref ComponentHandleState args)
+    {
+        if (args.Current is not NewGunComponentState state) return;
+
+        Sawmill.Debug($"Handle state: setting shot count from {component.ShotCounter} to {state.ShotCounter}");
+        component.NextFire = state.NextFire;
+        component.ShotCounter = state.ShotCounter;
+        component.FakeAmmo = state.FakeAmmo;
+        component.SelectedMode = state.SelectiveFire;
+        component.AvailableModes = state.AvailableSelectiveFire;
+    }
+
+    protected GunComponent? GetGun(EntityUid entity)
+    {
+        if (!EntityManager.TryGetComponent(entity, out SharedHandsComponent? hands) ||
+            hands.ActiveHandEntity is not { } held)
+        {
+            return null;
+        }
+
+        if (!EntityManager.TryGetComponent(held, out GunComponent? gun))
+            return null;
+
+        if (!_combatMode.IsInCombatMode(entity))
+            return null;
+
+        return gun;
+    }
+
+    private void StopShooting(GunComponent gun)
+    {
+        if (gun.ShotCounter == 0) return;
+
+        Sawmill.Debug($"Stopped shooting {ToPrettyString(gun.Owner)}");
+        gun.ShotCounter = 0;
+        gun.ShootCoordinates = null;
+        Dirty(gun);
+    }
+
+    private void AttemptShoot(EntityUid user, GunComponent gun)
+    {
+        if (gun.FireRate <= 0f) return;
+
+        var toCoordinates = gun.ShootCoordinates;
+
+        if (toCoordinates == null) return;
+
+        var curTime = Timing.CurTime;
+
+        // Need to do this to play the clicking sound for empty automatic weapons
+        // but not play anything for burst fire.
+        if (gun.NextFire > curTime) return;
+
+        // First shot
+        if (gun.ShotCounter == 0 && gun.NextFire < curTime)
+            gun.NextFire = curTime;
+
+        var shots = 0;
+        var lastFire = gun.NextFire;
+        var fireRate = TimeSpan.FromSeconds(1f / gun.FireRate);
+
+        while (gun.NextFire <= curTime)
+        {
+            gun.NextFire += fireRate;
+            shots++;
+        }
+
+        // Get how many shots we're actually allowed to make, due to clip size or otherwise.
+        // Don't do this in the loop so we still reset NextFire.
+        switch (gun.SelectedMode)
+        {
+            case SelectiveFire.SemiAuto:
+                shots = Math.Min(shots, 1 - gun.ShotCounter);
```

Comment:
Could this accidentally go negative, and trip that `DebugTools.Assert(shots >= 0);`  below? Not sure.

Reply:
It'll never hit here with fewer than 1 shot. Maybe if the selector mode is toggled while ShotCounter is active but in which case I'd want the assert tripping anyway.